62_CAD/shaders/main_pipeline/fragment_shader.hlsl

```diff
@@ -1,655 +1,1104 @@
+    }
+
+    float minDistance = nbl::hlsl::numeric_limits<float>::max;
+    if (!outlineStyle.hasStipples() || stretch == InvalidStyleStretchValue)
+    {
+        for (int i = 0; i < 3; ++i)
+        {
+            if (distances[i] > outlineThickness)
+                continue;
+
+            const uint2 currentEdgePoints = edgePoints[i];
+            float3 p0 = v[currentEdgePoints[0]];
+            float3 p1 = v[currentEdgePoints[1]];
+
+            float distance = nbl::hlsl::numeric_limits<float>::max;
+            nbl::hlsl::shapes::Line<float> lineSegment = nbl::hlsl::shapes::Line<float>::construct(float2(p0.x, p0.y), float2(p1.x, p1.y));
+            distance = ClippedSignedDistance<nbl::hlsl::shapes::Line<float> >::sdf(lineSegment, psInput.position.xy, outlineThickness, outlineStyle.isRoadStyleFlag);
+
+            minDistance = min(minDistance, distance);
+        }
+    }
+    else
+    {
+        for (int i = 0; i < 3; ++i)
+        {
+            if (distances[i] > outlineThickness)
+                continue;
+
+            const uint2 currentEdgePoints = edgePoints[i];
+            float3 p0 = v[currentEdgePoints[0]];
+            float3 p1 = v[currentEdgePoints[1]];
+
+            // long story short, in order for stipple patterns to be consistent:
+            // - point with lesser x coord should be starting point
+            // - if x coord of both points are equal then point with lesser y value should be starting point
+            if (p1.x < p0.x)
+                nbl::hlsl::swap(p0, p1);
+            else if (p1.x == p0.x && p1.y < p0.y)
+                nbl::hlsl::swap(p0, p1);
+
+            nbl::hlsl::shapes::Line<float> lineSegment = nbl::hlsl::shapes::Line<float>::construct(float2(p0.x, p0.y), float2(p1.x, p1.y));
+
+            float distance = nbl::hlsl::numeric_limits<float>::max;
+            nbl::hlsl::shapes::Line<float>::ArcLengthCalculator arcLenCalc = nbl::hlsl::shapes::Line<float>::ArcLengthCalculator::construct(lineSegment);
+            LineStyleClipper clipper = LineStyleClipper::construct(outlineStyle, lineSegment, arcLenCalc, phaseShift, stretch, worldToScreenRatio);
+            distance = ClippedSignedDistance<nbl::hlsl::shapes::Line<float>, LineStyleClipper>::sdf(lineSegment, psInput.position.xy, outlineThickness, outlineStyle.isRoadStyleFlag, clipper);
+
+            minDistance = min(minDistance, distance);
+        }
+
+    }
+
+    outputColor.a = smoothstep(+globals.antiAliasingFactor, -globals.antiAliasingFactor, minDistance) * outlineStyle.color.a;
+    outputColor.rgb = outlineStyle.color.rgb;
+
+    return outputColor;
+}
+
+float4 blendColorOnTop(in float4 colorBelow, in float4 colorAbove)
+{
+    float4 outputColor = colorBelow;
+    outputColor.rgb = colorAbove.rgb * colorAbove.a + outputColor.rgb * outputColor.a * (1.0f - colorAbove.a);
+    outputColor.a = colorAbove.a + outputColor.a * (1.0f - colorAbove.a);

```

Comment:
There are a few notes regarding this custom blending which I believe might be the cause of some bugs:

- first ever blendColorOnTop, should override outputColor (dtmColor) and not blend at all. you're wrongly blending with color black (let's say your only have height shading and it's color is (1, 0, 0, 0.5)), background color can be anything. you should let the hardware do the final blend.

- We either needs to max the alphas or replace with srcAlpha(colorAbove.a) completely. I need to think about what makes more sense.

- It's correct that we're using srcAlpha to blend dst and src color, but it only makes sense if we blend that after it's blended with the background.

TL;DR I need some time to figure this shit out

Reply:
Ok we should do under blending, change this function's name to `blendUnder` and then render the other way around.
and as an optimization skip rendering the next stuff if alpha is 1.0 (the color is gonna get discarded for things behind)

the under blending operator is as follows:
```
dst.rgb = dst.rgb*dst.a + (1-dst.a)*src.a*src.rgb; // src and dst swapped
dst.a = (1.0f-src.a)*dst.a + src.a; // like currently implemented
``
the current implementation while it looks correct, fails when you just blend a bunch of fully transparent things on top of a semi transparent thing because `outputColor.rgb = outputColor.rgb*outputColor.a;` each time.

Reply:
Edits, after putting some thought into it and reading [dual depth](https://developer.download.nvidia.com/SDK/10/opengl/src/dual_depth_peeling/doc/DualDepthPeeling.pdf), we should replace `dst.rgb*dst.a` with just `dst.rgb` because the multiplication was taken into account in previous steps by `src.a*src.rgb`

so the whole expression becomes:
```
dst.rgb = dst.rgb + (1-dst.a)*src.a*src.rgb; // src and dst swapped
dst.a = (1.0f-src.a)*dst.a + src.a; // like currently implemented
```

test if this fixes your issue [here](https://discord.com/channels/593902898015109131/1338454683869515847/1364345059109240943)
don't forget to reverse the drawing order.